File: Scripts/test_mcp_tools.py
#!/usr/bin/env python3
"""Test MCP tools using the FastMCP server via stdio."""

import json
import subprocess
import sys
from pathlib import Path

project_root = Path(__file__).parent.parent
venv_python = project_root / ".venv" / "bin" / "python"

class MCPClient:
    """Simple MCP client using stdio."""
    
    def __init__(self):
        self.process = None
        self.request_id = 0
        self._start_server()
        self._initialize()
    
    def _start_server(self):
        """Start the MCP server in stdio mode."""
        env = {}
        env["MCP_TRANSPORT"] = "stdio"
        
        self.process = subprocess.Popen(
            [str(venv_python), str(project_root / "server.py")],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
            cwd=str(project_root),
            env={**dict(subprocess.os.environ), **env}
        )
    
    def _send_jsonrpc(self, method: str, params=None):
        """Send a JSON-RPC 2.0 request and get the response."""
        self.request_id += 1
        
        payload = {
            "jsonrpc": "2.0",
            "id": self.request_id,
            "method": method,
        }
        if params:
            payload["params"] = params
        
        try:
            self.process.stdin.write(json.dumps(payload) + "\n")
            self.process.stdin.flush()
            
            response_line = self.process.stdout.readline()
            if not response_line:
                raise RuntimeError("Server closed connection")
            
            return json.loads(response_line)
        except Exception as e:
            raise RuntimeError(f"MCP communication error: {e}")
    
    def _initialize(self):
        """Initialize the MCP connection."""
        try:
            response = self._send_jsonrpc("initialize", {
                "protocolVersion": "2024-11-05",
                "capabilities": {},
                "clientInfo": {
                    "name": "test-mcp-tools-script",
                    "version": "1.0.0"
                }
            })
            if response.get("error"):
                raise RuntimeError(f"Initialize failed: {response['error']}")
        except Exception as e:
            self.close()
            raise
    
    def list_tools(self):
        """List available tools."""
        response = self._send_jsonrpc("tools/list")
        return response
    
    def call_tool(self, tool_name: str, arguments: dict):
        """Call a tool on the MCP server."""
        response = self._send_jsonrpc("tools/call", {
            "name": tool_name,
            "arguments": arguments
        })
        return response
    
    def close(self):
        """Close the server process."""
        if self.process:
            self.process.terminate()
            try:
                self.process.wait(timeout=2)
            except subprocess.TimeoutExpired:
                self.process.kill()

class MCPTester:
    """Test suite for MCP tools."""
    
    def __init__(self):
        self.client = None
    
    def start(self):
        """Initialize the client."""
        self.client = MCPClient()
    
    def test_list_tools(self) -> bool:
        """Test tools/list method."""
        print("📋 Testing tools/list...")
        
        try:
            response = self.client.list_tools()
            
            if response.get("error"):
                print(f"❌ tools/list failed: {response['error']}")
                return False
            
            tools = response.get("result", {}).get("tools", [])
            expected_tools = ["salesforce_query", "salesforce_sobjects", "salesforce_recent"]
            
            print(f"✅ Found {len(tools)} tools:")
            for tool in tools:
                print(f"   - {tool['name']}: {tool['description']}")
            
            has_all_tools = all(
                any(tool["name"] == expected for tool in tools)
                for expected in expected_tools
            )
            
            if has_all_tools:
                print("✅ All expected tools are available\n")
                return True
            else:
                print("❌ Some expected tools are missing\n")
                return False
        except Exception as e:
            print(f"❌ Error: {e}")
            return False
    
    def test_tool(self, tool_name: str, args=None) -> bool:
        """Test a specific tool."""
        if args is None:
            args = {}
        
        print(f"🔧 Testing {tool_name}...")
        
        try:
            response = self.client.call_tool(tool_name, args)
            
            if response.get("error"):
                print(f"❌ {tool_name} failed: {response['error']}")
                return False
            
            content = response.get("result", {}).get("content", [{}])[0].get("text", "")
            
            if content:
                try:
                    data = json.loads(content)
                    if isinstance(data, dict):
                        keys = list(data.keys())[:3]
                        print(f"✅ {tool_name} returned data: {type(data).__name__} with keys: {keys}")
                    elif isinstance(data, list):
                        print(f"✅ {tool_name} returned data: list with {len(data)} items")
                    else:
                        print(f"✅ {tool_name} returned data: {type(data).__name__}")
                except json.JSONDecodeError:
                    print(f"✅ {tool_name} returned text ({len(content)} chars)")
            else:
                print(f"⚠️  {tool_name} returned no content")
            
            return True
        except Exception as e:
            print(f"❌ Error: {e}")
            return False
    
    def run_all_tests(self):
        """Run all tests."""
        results = {
            "serverStart": False,
            "listTools": False,
            "query": False,
            "sobjects": False,
            "recent": False
        }
        
        try:
            print("🚀 Starting MCP server...\n")
            self.start()
            results["serverStart"] = True
            
            results["listTools"] = self.test_list_tools()
            results["query"] = self.test_tool("salesforce_query", {"q": "SELECT Id FROM Account LIMIT 1"})
            results["sobjects"] = self.test_tool("salesforce_sobjects")
            results["recent"] = self.test_tool("salesforce_recent", {"limit": 5})
            
        except Exception as e:
            print(f"Test suite error: {e}")
        finally:
            self.cleanup()
        
        return results
    
    def print_summary(self, results):
        """Print test results summary."""
        print("\n" + "=" * 50)
        print("📊 MCP SERVER TEST RESULTS")
        print("=" * 50)
        
        tests = [
            ("Server Startup", results["serverStart"]),
            ("List Tools", results["listTools"]),
            ("Salesforce Query", results["query"]),
            ("Salesforce SObjects", results["sobjects"]),
            ("Salesforce Recent", results["recent"])
        ]
        
        for name, passed in tests:
            status = "✅" if passed else "❌"
            print(f"{status} {name}")
        
        passed_count = sum(1 for v in results.values() if v)
        total_count = len(results)
        
        print(f"\n📈 Summary: {passed_count}/{total_count} tests passed")
        
        if passed_count == total_count:
            print("🎉 All tests passed! Your Salesforce MCP connector is working perfectly!")
        else:
            print("⚠️  Some tests failed. Check the logs above for details.")
    
    def cleanup(self):
        """Clean up resources."""
        if self.client:
            self.client.close()

def main():
    """Main function."""
    tester = MCPTester()
    results = tester.run_all_tests()
    tester.print_summary(results)
    
    sys.exit(0 if all(results.values()) else 1)

if __name__ == "__main__":
    main()

File: Scripts/test-mcp-tools.js
#!/usr/bin/env node

import { spawn } from 'child_process';
import { setTimeout } from 'timers/promises';

class MCPTester {
  constructor() {
    this.server = null;
    this.requestId = 1;
  }

  async startServer() {
    console.log('🚀 Starting MCP server...');
    
    this.server = spawn('node', ['server.js'], {
      stdio: ['pipe', 'pipe', 'pipe']
    });

    let serverReady = false;
    
    this.server.stderr.on('data', (data) => {
      const output = data.toString();
      if (output.includes('Salesforce MCP server running on stdio')) {
        serverReady = true;
      }
    });

    // Wait for server to be ready
    let attempts = 0;
    while (!serverReady && attempts < 10) {
      await setTimeout(200);
      attempts++;
    }

    if (!serverReady) {
      throw new Error('Server failed to start within timeout');
    }

    console.log('✅ MCP server started successfully\n');
    return true;
  }

  async sendMCPRequest(method, params = {}) {
    const request = {
      jsonrpc: '2.0',
      id: this.requestId++,
      method,
      params
    };

    const requestStr = JSON.stringify(request) + '\n';
    
    return new Promise((resolve, reject) => {
      let responseData = '';
      let timeoutId;

      const onData = (data) => {
        responseData += data.toString();
        
        // Look for complete JSON response
        try {
          const lines = responseData.split('\n').filter(line => line.trim());
          for (const line of lines) {
            const response = JSON.parse(line);
            if (response.id === request.id) {
              clearTimeout(timeoutId);
              this.server.stdout.off('data', onData);
              resolve(response);
              return;
            }
          }
        } catch (e) {
          // Continue collecting data
        }
      };

      this.server.stdout.on('data', onData);
      
      // Set timeout
      timeoutId = global.setTimeout(() => {
        this.server.stdout.off('data', onData);
        reject(new Error(`Request timeout for method: ${method}`));
      }, 10000);

      // Send request
      this.server.stdin.write(requestStr);
    });
  }

  async testListTools() {
    console.log('📋 Testing tools/list...');
    
    try {
      const response = await this.sendMCPRequest('tools/list');
      
      if (response.error) {
        console.log('❌ tools/list failed:', response.error.message);
        return false;
      }

      const tools = response.result?.tools || [];
      const expectedTools = ['salesforce_query', 'salesforce_sobjects', 'salesforce_recent'];
      
      console.log(`✅ Found ${tools.length} tools:`);
      tools.forEach(tool => {
        console.log(`   - ${tool.name}: ${tool.description}`);
      });

      const hasAllTools = expectedTools.every(expectedTool => 
        tools.some(tool => tool.name === expectedTool)
      );

      if (hasAllTools) {
        console.log('✅ All expected tools are available\n');
        return true;
      } else {
        console.log('❌ Some expected tools are missing\n');
        return false;
      }
    } catch (error) {
      console.log('❌ tools/list error:', error.message);
      return false;
    }
  }

  async testTool(toolName, args = {}) {
    console.log(`🔧 Testing ${toolName}...`);
    
    try {
      const response = await this.sendMCPRequest('tools/call', {
        name: toolName,
        arguments: args
      });

      if (response.error) {
        console.log(`❌ ${toolName} failed:`, response.error.message);
        return false;
      }

      const content = response.result?.content?.[0]?.text;
      if (content) {
        try {
          const data = JSON.parse(content);
          console.log(`✅ ${toolName} returned data:`, typeof data, Object.keys(data).slice(0, 3));
        } catch (e) {
          console.log(`✅ ${toolName} returned text (${content.length} chars)`);
        }
      } else {
        console.log(`⚠️  ${toolName} returned no content`);
      }
      
      return true;
    } catch (error) {
      console.log(`❌ ${toolName} error:`, error.message);
      return false;
    }
  }

  async runAllTests() {
    const results = {
      serverStart: false,
      listTools: false,
      query: false,
      sobjects: false,
      recent: false
    };

    try {
      // Start server
      results.serverStart = await this.startServer();

      // Test list tools
      results.listTools = await this.testListTools();

      // Test individual tools
      results.query = await this.testTool('salesforce_query', { 
        q: 'SELECT Id FROM Account LIMIT 1' 
      });
      
      results.sobjects = await this.testTool('salesforce_sobjects');
      
      results.recent = await this.testTool('salesforce_recent', { 
        limit: 5 
      });

    } catch (error) {
      console.error('Test suite error:', error);
    } finally {
      // Clean up
      if (this.server) {
        this.server.kill('SIGTERM');
      }
    }

    return results;
  }

  printSummary(results) {
    console.log('\n' + '='.repeat(50));
    console.log('📊 MCP SERVER TEST RESULTS');
    console.log('='.repeat(50));
    
    const tests = [
      ['Server Startup', results.serverStart],
      ['List Tools', results.listTools], 
      ['Salesforce Query', results.query],
      ['Salesforce SObjects', results.sobjects],
      ['Salesforce Recent', results.recent]
    ];

    tests.forEach(([name, passed]) => {
      console.log(`${passed ? '✅' : '❌'} ${name}`);
    });

    const passedCount = Object.values(results).filter(Boolean).length;
    const totalCount = Object.keys(results).length;
    
    console.log(`\n📈 Summary: ${passedCount}/${totalCount} tests passed`);
    
    if (passedCount === totalCount) {
      console.log('🎉 All tests passed! Your Salesforce MCP connector is working perfectly!');
    } else {
      console.log('⚠️  Some tests failed. Check the logs above for details.');
    }
  }
}

// Run tests
async function main() {
  const tester = new MCPTester();
  const results = await tester.runAllTests();
  tester.printSummary(results);
  
  process.exit(Object.values(results).every(Boolean) ? 0 : 1);
}

main().catch(console.error);